File: src/harness/bridge/protocol.py
"""Immutable version 1 Browser Bridge protocol values."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from enum import Enum

from harness.agent.values import JsonValue, freeze_json, freeze_json_object

PROTOCOL_VERSION = "1"


class PagePluginState(str, Enum):
    """Page-local state for one exact client revision."""

    ABSENT = "absent"
    LOADING = "loading"
    ACTIVE = "active"
    WAITING = "waiting"
    FAILED = "failed"
    UNLOADING = "unloading"


@dataclass(frozen=True, slots=True)
class DesiredClient:
    """One client revision desired by the Host."""

    plugin_id: str
    revision: str
    bundle_url: str
    bundle_sha256: str


@dataclass(frozen=True, slots=True)
class ReconcileCommand:
    """Complete desired client graph for one page operation."""

    protocol: str
    operation_id: str
    desired: tuple[DesiredClient, ...]

    def __post_init__(self) -> None:
        if self.protocol != PROTOCOL_VERSION:
            raise ValueError("unsupported Browser Bridge protocol")
        object.__setattr__(self, "desired", tuple(self.desired))


@dataclass(frozen=True, slots=True)
class PluginLoadResult:
    """Page result for one plugin in a reconciliation operation."""

    protocol: str
    operation_id: str
    plugin_id: str
    revision: str
    state: PagePluginState
    error: str | None = None


@dataclass(frozen=True, slots=True)
class RpcCall:
    """Package-private browser-to-backend call."""

    protocol: str
    page_id: str
    call_id: str
    plugin_id: str
    revision: str
    method: str
    arguments: Mapping[str, JsonValue]

    def __post_init__(self) -> None:
        if self.protocol != PROTOCOL_VERSION:
            raise ValueError("unsupported Browser Bridge protocol")
        object.__setattr__(self, "arguments", freeze_json_object(self.arguments))


@dataclass(frozen=True, slots=True)
class RpcResult:
    """Structured RPC success or failure response."""

    protocol: str
    call_id: str
    result: JsonValue | None = None
    error_code: str | None = None
    error_message: str | None = None

    def __post_init__(self) -> None:
        if self.protocol != PROTOCOL_VERSION:
            raise ValueError("unsupported Browser Bridge protocol")
        if self.error_code is None:
            object.__setattr__(self, "result", freeze_json(self.result))
        elif self.result is not None or not self.error_message:
            raise ValueError("failed RPC result requires only code and message")

File: frontend/src/protocol.ts
/** JSON values admitted by the Browser Bridge wire protocol. */
export type JsonValue = null | boolean | number | string | JsonValue[] | { [key: string]: JsonValue }

/** One exact client contribution desired by the Python Host. */
export interface DesiredClient {
  pluginId: string
  revision: string
  bundleUrl: string
  bundleSha256: string
  protocolSchemaUrl: string | null
  activationPolicy: 'required' | 'optional'
}

/** Complete desired graph for one superseding operation. */
export interface ReconcileCommand {
  protocol: '1'
  type: 'reconcile'
  operationId: string
  desired: DesiredClient[]
}

/** Page-local lifecycle state reported for one exact Revision. */
export type PagePluginState = 'absent' | 'loading' | 'active' | 'waiting' | 'failed' | 'unloading'

/** One page-local plugin result. */
export interface PluginLoadResult {
  protocol: '1'
  type: 'plugin-result'
  operationId: string
  pluginId: string
  revision: string
  state: PagePluginState
  error: string | null
}

/** Terminal result for one reconciliation operation. */
export interface ReconcileComplete {
  protocol: '1'
  type: 'reconcile-complete'
  operationId: string
  success: boolean
  error: string | null
}

/** Frames emitted by the Cordis client adapter during reconciliation. */
export type ReconcileResultFrame = PluginLoadResult | ReconcileComplete

/** Initial inventory sent once for a logical browser connection. */
export interface HelloFrame {
  protocol: '1'
  type: 'hello'
  pageId: string
  loaded: Record<string, string>
}

/** Same-plugin/revision call sent from a client contribution. */
export interface RpcCallFrame {
  protocol: '1'
  type: 'rpc-call'
  pageId: string
  callId: string
  pluginId: string
  revision: string
  method: string
  arguments: Record<string, JsonValue>
}

/** Structured backend result for one RPC call. */
export type RpcResultFrame = {
  protocol: '1'
  type: 'rpc-result'
  callId: string
  result: JsonValue
} | {
  protocol: '1'
  type: 'rpc-result'
  callId: string
  errorCode: string
  errorMessage: string
}

/** Best-effort cancellation for one active RPC call. */
export interface RpcCancelFrame {
  protocol: '1'
  type: 'rpc-cancel'
  pageId: string
  callId: string
}

/** Explicit same-plugin/revision Event in either direction. */
export interface BridgeEventFrame {
  protocol: '1'
  type: 'event'
  pageId: string
  pluginId: string
  revision: string
  name: string
  payload: JsonValue
}

/** Browser-originated frames emitted by the client runtime. */
export type ClientFrame = HelloFrame | ReconcileResultFrame | RpcCallFrame | RpcCancelFrame | BridgeEventFrame

/** Host-originated frames accepted by the client runtime. */
export type ServerFrame = ReconcileCommand | RpcResultFrame | BridgeEventFrame
